internal/pkg/consensus/block_validation.go

```diff
@@ -104,6 +126,81 @@ func (dv *DefaultBlockValidator) ValidateSemantic(ctx context.Context, child *bl
 	return nil
 }
 
+// ValidateFullSemantic checks validation conditions on a block's messages that don't require message execution.
+func (dv *DefaultBlockValidator) ValidateMessagesSemantic(ctx context.Context, child *block.Block, parents block.TipSetKey) error {
+	// validate call sequence numbers
+	secpMsgs, blsMsgs, err := dv.ms.LoadMessages(ctx, child.Messages.Cid)
+	if err != nil {
+		return errors.Wrapf(err, "block validation failed loading message list %s for block %s", child.Messages, child.Cid())
+	}
+
+	expectedCallSeqNum := map[address.Address]uint64{}
+	for _, msg := range blsMsgs {
+		msgCid, err := msg.Cid()
+		if err != nil {
+			return err
+		}
+
+		from, err := dv.getAndValidateFromActor(ctx, msg, parents)
+		if err != nil {
+			return errors.Wrapf(err, "from actor %s for message %s of block %s invalid", msg.From, msgCid, child.Cid())
+		}
+
+		err = dv.validateMessage(msg, expectedCallSeqNum, from)
+		if err != nil {
+			return errors.Wrapf(err, "message %s of block %s invalid", msgCid, child.Cid())
+		}
+	}
+
+	for _, msg := range secpMsgs {
+		msgCid, err := msg.Cid()
+		if err != nil {
+			return err
+		}
+
+		from, err := dv.getAndValidateFromActor(ctx, &msg.Message, parents)
+		if err != nil {
+			return errors.Wrapf(err, "from actor %s for message %s of block %s invalid", msg.Message.From, msgCid, child.Cid())
+		}
+
+		err = dv.validateMessage(&msg.Message, expectedCallSeqNum, from)
+		if err != nil {
+			return errors.Wrapf(err, "message %s of block %s invalid", msgCid, child.Cid())
+		}
+	}
+
+	return nil
+}
+
+func (dv *DefaultBlockValidator) getAndValidateFromActor(ctx context.Context, msg *types.UnsignedMessage, parents block.TipSetKey) (*actor.Actor, error) {
+	actor, err := dv.cs.GetActorAt(ctx, parents, msg.From)
```

Comment:
This needs to resolve pubkey addresses to ID addresses via the init actor state, which I don't think the underlying chain store does. It may be worth doing state access here via state view methods, which could do the resolution.

Reply:
`GetActorAt` actually does the resolution internally, and finds it at the tipset specified by parents. It has to since actors are keyed in the state tree by ID address. I've confirmed that this code is covered by integration tests that fetch actual actors.

I'm wondering if we should create an actor IDAddress type definition in GFC (or possibly specs-actors) that would signal when an address is not going to get resolved. I foresee a lot more bugs and confusion around this.